### PR TITLE
Add encapsulation reference doc

### PR DESCRIPTION
**Pull request by jsumners**
We keep having to answer questions about Fastify's encapsulation feature because we do not currently explain it thoroughly in our reference documentation. This PR solves that issue.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Encapsulation.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    * One _grandchild context_ where each _grandchild context_ has
+        - Three _child plugins_
+
+Every _child context_ and _grandchild context_ has access to the _root plugins_.
+Within each _child context_ the _granchild contexts_ have access to the
+_child plugins_ registered within the containing _child context_, but the
+containing _child context_ **does not** have access to the _child plugins_
+registered within its _grandchild context_.
+
+Given that everything in Fastify is a [plugin](./Plugins.md), exept for the
```
Comment (jsumners) [inline docs/Encapsulation.md:29]:
```suggestion
Given that everything in Fastify is a [plugin](./Plugins.md), except for the
```

Comment (kibertoad) [inline docs/Encapsulation.md:29]: "except"

File: docs/Encapsulation.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+_root context_, every "context" and "plugin" in this example is a plugin
+that can consist of decorators, hooks, plugins, and routes. Thus, to put
+this example into concrete terms, let's consider a basic scenario of a REST API
```
Comment (jsumners) [inline docs/Encapsulation.md:32]:
```suggestion
this example into concrete terms, consider a basic scenario of a REST API
```

File: docs/Encapsulation.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+fastify.listen(8000)
+```
+
+The above server example shows all of the encapsulation concepts outlined in the
+original diagram:
+
+1. Each _child context_, `authenticatedContext`, `publicContext`, and
+`grandchildContext` have access to the `answer` request decorator defined in
+the _root context.
```
Comment (jsumners) [inline docs/Encapsulation.md:102]:
```suggestion
1. Each _child context_ (`authenticatedContext`, `publicContext`, and
`grandchildContext`) has access to the `answer` request decorator defined in
the _root context_.
```

File: docs/Encapsulation.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+{"answer":42,"foo":"foo","bar":"bar"}
+```
+
+[bearer]: https://github.com/fastify/fastify-bearer-auth
+
+<a id="shared-context"></a>
+## Sharing Between Contexts
+
+Notice that each context in the prior example inherits _only_ from the parent
+contexts. Sibibling contexts do not share any entities. This default is
```
Comment (kibertoad) [inline docs/Encapsulation.md:126]: "Sibling"?

File: docs/Encapsulation.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+occassionally not desired. In such cases, the encapsulation context can be
+broken through the usage of [fastify-plugin][fastify-plugin] such that
+anything registered in a descendent context is availble to the containing
```
Comment (kibertoad) [inline docs/Encapsulation.md:129]: "available"

File: docs/Encapsulation.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+parent context.
+
+Assuming the `publicContext` needs access to the `bar` decorator defined
+within the `grandchildContext` in the previous example, the code can be
+refactored to:
```
Comment (kibertoad) [inline docs/Encapsulation.md:134]: "Refactoring" implies zero change in the observable behaviour, which I'm not sure in this case.

Comment (jsumners) [reply docs/Encapsulation.md:134]:
```suggestion
rewritten to:
```

File: docs/Encapsulation.md (hunk 5 above)
Comment (jsumners) [reply docs/Encapsulation.md:129]:
```suggestion
anything registered in a descendent context is available to the containing
```

File: docs/Encapsulation.md (hunk 4 above)
Comment (jsumners) [reply docs/Encapsulation.md:126]:
```suggestion
contexts. Sibling contexts do not share any entities. This default is
```

**Review by kibertoad — APPROVED** [review]

File: docs/Encapsulation.md (hunk 6 above)
Comment (jsumners) [reply docs/Encapsulation.md]: Changed to "rewritten" (as I originally typed but, uh, refactored 🤣).

**Review by zekth — APPROVED** [review]
lgtm. Good way to explain how to properly separate authenticated / non authenticated routes

File: docs/resources/encapsulation_context.svg [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (StarpTech) [inline docs/resources/encapsulation_context.svg:1]: Typo in "Grandchild contex"

Comment (jsumners) [reply docs/resources/encapsulation_context.svg:1]: https://github.com/fastify/fastify/pull/2725

**Comment by kibertoad** [comment]
@jsumners BTW, new article does not show up on https://www.fastify.io/docs/latest/
I assume TOC needs to be updated as well?

**Comment by jsumners** [comment]
@kibertoad I put it in `/README.md`. 

@lmammino how does this file get included?

**Pull request merged**